### PR TITLE
Feature: Added option to show some features outside SB.

### PR DESCRIPTION
**Pull request by Obsidianninja11**
Hopefully it works this time
Idk what happened when i tried to fix merge conflicts last time, but I couldn't figure out how to revert that

## Changelog Improvements
+ Added option to show some features outside SkyBlock. - Obsidian

### CODE REVIEW

**Review by CalMWolfs — CHANGES_REQUESTED** [review]
You have changed multiple config options location and names which means the user's settings for these specific settings will be reset.
Please look at the code base to see examples of ConfigUpdaterMigrator.ConfigFixEvent to correct this.

**Comment by Obsidianninja11** [comment]
I think i fixed everything.
Didn't test though (idrk how to)

File: src/main/java/at/hannibal2/skyhanni/config/features/GardenConfig.java [hunk 1]
```diff
@@ -799,6 +799,16 @@ public static class EliteFarmingWeightConfig {
...
         @ConfigEditorBoolean
         public boolean ignoreLow = false;
+
+        @Expose
+        @ConfigOption(name = "Show outside garden", desc = "Show the farming weight outside of the garden.")
+        @ConfigEditorBoolean
+        public boolean showOutsideGarden = false;
+
+        @Expose
+        @ConfigOption(name = "Show outside SB", desc = "Show the farming weight outside of SkyBlock.")
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/GardenConfig.java]: Why? Im sure we don't want this, also currently all the data gets reset on world switch anyway.

File: src/main/java/at/hannibal2/skyhanni/features/chat/playerchat/PlayerChatModifier.kt [hunk 2]
```diff
@@ -80,7 +80,7 @@ class PlayerChatModifier {
 
         if (SkyHanniMod.feature.markedPlayers.highlightInChat) {
             for (markedPlayer in MarkedPlayerManager.playerNamesToMark) {
-                string = string.replace(markedPlayer, "§e$markedPlayer§r")
+                string = string.replace(markedPlayer, "§e$markedPlayer§r", true)
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/chat/playerchat/PlayerChatModifier.kt]: please remove this from this pr and move it into a standalone pr

**Review by hannibal002 — CHANGES_REQUESTED** [review]

File: src/main/java/at/hannibal2/skyhanni/features/cosmetics/CosmeticFollowingLine.kt [hunk 3]
```diff
@@ -33,7 +33,7 @@ class CosmeticFollowingLine {
 
     @SubscribeEvent
     fun onRenderWorld(event: LorenzRenderWorldEvent) {
-        if (!LorenzUtils.inSkyBlock) return
+        if (!LorenzUtils.inSkyBlock && !config.showOutsideSB) return
```
Comment (walkerselby) [inline src/main/java/at/hannibal2/skyhanni/features/cosmetics/CosmeticFollowingLine.kt:36]:
```suggestion
        if (!LorenzUtils.inSkyBlock || config.showOutsideSB) return
```

File: src/main/java/at/hannibal2/skyhanni/features/chat/playerchat/PlayerChatModifier.kt (hunk 2 above)
Comment (Obsidianninja11) [reply src/main/java/at/hannibal2/skyhanni/features/chat/playerchat/PlayerChatModifier.kt]: oops. idk why thats there

File: src/main/java/at/hannibal2/skyhanni/config/features/GardenConfig.java (hunk 1 above)
Comment (Obsidianninja11) [reply src/main/java/at/hannibal2/skyhanni/config/features/GardenConfig.java]: I want it
Also the data didn't get reset on world switch? Leaderboard just didn't update outside of sb before.
I've fixed that now though. It'll use the selected profile from the elite api call (no new api call) when outside of sb.

**Comment by Obsidianninja11** [comment]
FINALLY IT WORKED

File: src/main/java/at/hannibal2/skyhanni/config/features/GUIConfig.java [hunk 4]
```diff
@@ -128,15 +130,14 @@ public static class InGameDateConfig {
...
     }
 
-
     @Expose
-    @ConfigOption(name = "TPS Display", desc = "Show the TPS of the current server, like in Soopy.")
+    @ConfigOption(name = "Enabled", desc = "Show the TPS of the current server, like in Soopy.")
     @ConfigEditorBoolean
     @FeatureToggle
-    public boolean tpsDisplay = false;
+    public boolean enabled = false;
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/GUIConfig.java]: DId you forgot to create a category for tps display and position?

**Review by hannibal002 — CHANGES_REQUESTED** [review]

File: src/main/java/at/hannibal2/skyhanni/config/features/GUIConfig.java (hunk 4 above)
Comment (Obsidianninja11) [reply src/main/java/at/hannibal2/skyhanni/config/features/GUIConfig.java]: I thought I removed the category for it.
I'll check tonight when I get back home

File: src/main/java/at/hannibal2/skyhanni/features/garden/farming/FarmingWeightDisplay.kt [hunk 5]
```diff
@@ -150,10 +150,6 @@ class FarmingWeightDisplay {
             if (weight == -1.0) {
                 if (!isLoadingWeight) {
                     val localProfile = HypixelData.profileName
-                    if (localProfile == "") {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/garden/farming/FarmingWeightDisplay.kt:156]: Why remove this? I have added this since some users were having crashes without.

File: src/main/java/at/hannibal2/skyhanni/features/garden/farming/FarmingWeightDisplay.kt [hunk 6]
```diff
@@ -137,7 +137,7 @@ class FarmingWeightDisplay {
         }
 
         private fun update() {
-            if (!GardenAPI.inGarden()) return
+            if (!isEnabled()) return
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/garden/farming/FarmingWeightDisplay.kt:148]: Please make it so that it only updates once when outside garden. Moreover, you should handle world switches differently. No need to update the weight display on every world switch when the user wants to see the weight outside SkyBlock. Also the user will most likely ask for an option to show the display outside the garden but not outside SB. maybe we should add that as well? Personally, I don't think this whole feature is needed for the farming weight display at all, though.

**Review by hannibal002 — CHANGES_REQUESTED** [review]
I will move this PR to 0.22 as the current logic is highly unreadable. I will add enums shortly and also provide a helper extension function to combat those long lines, but the core logic is still too complex to dump that code duplication on everything. We might want to use a function like "is in SB or has this toggle enabled" but there are too many edge cases for me to add that on the fly.

File: src/main/java/at/hannibal2/skyhanni/features/garden/farming/FarmingWeightDisplay.kt (hunk 5 above)
Comment (Obsidianninja11) [reply src/main/java/at/hannibal2/skyhanni/features/garden/farming/FarmingWeightDisplay.kt:156]: localProfile is blank when first logging into hypixel
and if the localProfile is blank, it will show the weight for the selected profile according to the elite api

File: src/main/java/at/hannibal2/skyhanni/features/garden/farming/FarmingWeightDisplay.kt (hunk 6 above)
Comment (Obsidianninja11) [reply src/main/java/at/hannibal2/skyhanni/features/garden/farming/FarmingWeightDisplay.kt:148]: So I should return if the player is outside of sb and the weight is valid in the onTick event?
And return in the onWorldChange event if the player is outside of sb?
There's already an option to show it outside the garden but not sb. Also, people who pretty much only farm (like me), might always want to be able to see their weight. Even outside of sb

Comment (hannibal002) [reply src/main/java/at/hannibal2/skyhanni/features/garden/farming/FarmingWeightDisplay.kt:148]: > So I should return if the player is outside of sb and the weight is valid in the onTick event?
> And return in the onWorldChange event if the player is outside of sb?

Both yes.
Just watch out if the player is in SB at all during the world change event. Better change it entirely to the island change event, as this will already only work while inside SkyBlock (and don't get called multiple times per world switch)

**Comment by hannibal002** [comment]
Please note, the data is not loaded for farming weight and compost empty time when not joined SkyBlock before.

First join:
image

after joining SkyBlock:
image

I don't know if this is worth fixing this as you need to handle the whole data load logic manually.

**Comment by Obsidianninja11** [comment]
ok that was a bit scuffed, but i think it's fine

File: src/main/java/at/hannibal2/skyhanni/features/garden/farming/FarmingWeightDisplay.kt (hunk 6 above)
Comment (Obsidianninja11) [reply src/main/java/at/hannibal2/skyhanni/features/garden/farming/FarmingWeightDisplay.kt:148]: I'm not sure how to do this. It's more complicated than just doing this:
> So I should return if the player is outside of sb and the weight is valid in the onTick event?
And return in the onWorldChange event if the player is outside of sb?

File: src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java [hunk 7]
```diff
@@ -2,15 +2,26 @@
...
 import io.github.moulberry.moulconfig.annotations.ConfigOption;
 
+import java.util.ArrayList;
+import java.util.List;
+import io.github.moulberry.moulconfig.annotations.ConfigEditorDraggableList;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java]: please ensure you use the optimise imports setting in intellij as there are 5 unneeded imports in this file alone

File: src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java [hunk 8]
```diff
@@ -158,6 +174,12 @@ public class MiscConfig {
     @Expose
     public Position playerMovementSpeedPos = new Position(394, 124, false, true);
 
+    @Expose
+    @ConfigOption(name = "Pet Candy Used", desc = "Show the number of Pet Candy used on a pet.")
+    @ConfigEditorBoolean
+    @FeatureToggle
+    public boolean petCandyUsed = true;
+
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java:183]: not sure why this is here. Incorrect merge?

File: src/main/java/at/hannibal2/skyhanni/features/garden/GardenYawAndPitch.kt [hunk 9]
```diff
@@ -18,10 +20,9 @@ class GardenYawAndPitch {
 
     @SubscribeEvent
     fun onRenderOverlay(event: GuiRenderEvent.GuiOverlayRenderEvent) {
-        if (!LorenzUtils.inSkyBlock) return
-        if (!config.enabled) return
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/garden/GardenYawAndPitch.kt:23]: this line shouldnt be removed

**Review by CalMWolfs — CHANGES_REQUESTED** [review]
need to change these things 
also did not check if all of your config fixes and isEnabled() statements were correct

File: src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java (hunk 8 above)
Comment (Obsidianninja11) [reply src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java:183]: I guess

File: src/main/java/at/hannibal2/skyhanni/features/garden/GardenYawAndPitch.kt (hunk 9 above)
Comment (Obsidianninja11) [reply src/main/java/at/hannibal2/skyhanni/features/garden/GardenYawAndPitch.kt:23]: Yes it should.
The checks were changed. Look at the lines below it.

File: src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java [hunk 10]
```diff
@@ -165,6 +175,12 @@ public class MiscConfig {
...
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java:183]: Why is this here

File: src/main/java/at/hannibal2/skyhanni/features/garden/GardenNextJacobContest.kt [hunk 11]
```diff
@@ -508,8 +509,9 @@ object GardenNextJacobContest {
         }
     }
 
-    private fun isEnabled() = LorenzUtils.inSkyBlock && config.display
-        && (GardenAPI.inGarden() || config.everywhere)
+    private fun isEnabled() = ((OutsideSbFeature.NEXT_JACOB_CONTEXT.isSelected() && !LorenzUtils.inSkyBlock) ||
+        (LorenzUtils.inSkyBlock && (GardenAPI.inGarden() || config.showOutsideGarden)))
+        && config.display
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/garden/GardenNextJacobContest.kt]: maybe something more like this idk


    private fun isEnabled() = config.display && (OutsideSbFeature.NEXT_JACOB_CONTEXT.isSelected() ||
    GardenAPI.inGarden() || (LorenzUtils.inSkyBlock && config.showOutsideGarden))

File: src/main/java/at/hannibal2/skyhanni/features/garden/GardenYawAndPitch.kt
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/garden/GardenYawAndPitch.kt:1]: You no longer check for config.enabled

File: src/main/java/at/hannibal2/skyhanni/features/garden/GardenYawAndPitch.kt [hunk 12]
```diff
@@ -57,4 +58,14 @@ class GardenYawAndPitch {
     fun onGardenToolChange(event: GardenToolChangeEvent) {
         lastChange = SimpleTimeMark.farPast()
     }
+
+    @SubscribeEvent
+    fun onConfigFix(event: ConfigUpdaterMigrator.ConfigFixEvent) {
+        event.move(18, "garden.yawPitchDisplay.everywhere", "garden.yawPitchDisplay.showOutsideGarden")
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/garden/GardenYawAndPitch.kt]: this isnt the correct previous config variable name

File: src/main/java/at/hannibal2/skyhanni/features/garden/GardenYawAndPitch.kt [hunk 13]
```diff
@@ -57,4 +58,14 @@ class GardenYawAndPitch {
...
+    }
+
+    private fun isEnabled() = (
+            (OutsideSbFeature.YAW_AND_PITCH.isSelected() && !LorenzUtils.inSkyBlock) ||
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/garden/GardenYawAndPitch.kt]: why do you check if not in skyblock, seems like a redundant check

File: src/main/java/at/hannibal2/skyhanni/features/garden/composter/ComposterDisplay.kt [hunk 14]
```diff
@@ -167,8 +168,10 @@ class ComposterDisplay {
             }
         } else "?"
 
-        if (!GardenAPI.inGarden() && config.displayOutsideGarden) {
-            val list = Collections.singletonList(listOf(NEUItems.getItemStack("BUCKET"), "§b$format"))
+        val inSb = LorenzUtils.inSkyBlock && config.displayOutsideGarden
+        val outsideSb = !LorenzUtils.inSkyBlock && OutsideSbFeature.COMPOSTER_TIME.isSelected()
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/garden/composter/ComposterDisplay.kt:172]: again, why even check if outside sb

File: src/main/java/at/hannibal2/skyhanni/features/misc/visualwords/ModifyVisualWords.kt [hunk 15]
```diff
@@ -17,7 +18,7 @@ object ModifyVisualWords {
         var modifiedText = originalText ?: return null
         if (!LorenzUtils.onHypixel) return originalText
         if (!config.enabled) return originalText
-        if (!LorenzUtils.inSkyBlock && !config.workOutside) return originalText
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/misc/visualwords/ModifyVisualWords.kt:20]: You never deleted the original config option for this

**Review by CalMWolfs — CHANGES_REQUESTED** [review]
overall still many things to change plus hannibal's comment to address,
Also I oppose Party Member Outlines outside of skyblock as this can easily give unfair advantage over other teams and be seen as a cheat

File: src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java (hunk 10 above)
Comment (Obsidianninja11) [reply src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java:183]: it was already there.
idk what happened. I didn't change it in last commit

File: src/main/java/at/hannibal2/skyhanni/features/garden/GardenYawAndPitch.kt (hunk 13 above)
Comment (Obsidianninja11) [reply src/main/java/at/hannibal2/skyhanni/features/garden/GardenYawAndPitch.kt]: bc outside sb doesn't make it show outside garden (like in hub)

File: src/main/java/at/hannibal2/skyhanni/features/garden/GardenNextJacobContest.kt (hunk 11 above)
Comment (Obsidianninja11) [reply src/main/java/at/hannibal2/skyhanni/features/garden/GardenNextJacobContest.kt]: No that would cause show outside sb to also make it show outside garden (e.g. in crimson isles)

**Comment by Obsidianninja11** [comment]
> Please make it so that it only updates once when outside garden. Moreover, you should handle world switches differently. No need to update the weight display on every world switch when the user wants to see the weight outside SkyBlock.

I haven't added this yet

**Comment by hannibal002** [comment]
I  want to fix the last thing missing in this pr myself: showing "join skyblock" in the lore for the 1-2 features that dont load properly in hypixel lobby

**Comment by Obsidianninja11** [comment]
> I want to fix the last thing missing in this pr myself: showing "join skyblock" in the lore for the 1-2 features that dont load properly in hypixel lobby

Oh yeah I forgot about that.

Also, do you still want the weight display to only update once when outside of sb?

**Comment by hannibal002** [comment]
> Also, do you still want the weight display to only update once when outside of sb?

yes, definetly.
there is no reason to keep fetching elite bot just to see how many spots you have falled since last skyblock join.

**Comment by Obsidianninja11** [comment]
> Also I oppose Party Member Outlines outside of skyblock as this can easily give unfair advantage over other teams and be seen as a cheat

Oh i forgot to remove this

**Comment by Obsidianninja11** [comment]
Oh and I don't think there's a config migration for the shwords show outside sb boolean to the draggable list

**Review by hannibal002 — APPROVED** [review]

File: src/main/java/at/hannibal2/skyhanni/features/garden/farming/FarmingWeightDisplay.kt (hunk 6 above)
Comment (hannibal002) [reply src/main/java/at/hannibal2/skyhanni/features/garden/farming/FarmingWeightDisplay.kt:148]: Both are no longer necessary, made it so it will normally update outside SkyBlock as inside SkyBlock.
